### ISSUE TITLE
Worker Queueing of Async Commands

### ISSUE THREAD

**Issue by iamtrask** [OpenMined/PySyft#1811]
This is a sub-ticket of Epic #1659.

This ticket is blocked by #1809.

In the previous issue (#1809), we created functionality which allowed for .send commands be sent asynchronously, returning a promise. However, promises cannot be used until the .wait() function guarantees that they are fulfilled. Trying to use a promise without calling .wait() throws an error.

In this ticket, we want to build on this functionality by eliminating the need for explicitly calling .wait() on promises. Instead, we want:

- [ ] 1) each ASYNC command also sends an ID (or several) which the resulting tensors should have.
- [ ] 2) each ASYNC command automatically returns pointers to the (yet to be created) results, using the IDs from (1)
- [ ] 3) If a pointer (on the client) is used before the object it points to has been created, the worker should put the operation into a queue, which waits until the tensor is created/available before executing.

**Comment by niuguy** [comment]
I think your solution is elegant by making async tasks a chain,  however, have you ever checked this paper? http://papers.nips.cc/paper/4687-large-scale-distributed-deep-networks.pdf   I just had a glance,  it might be more practical for large scale , though you may say it's not real asyn...

>  To apply SGD to large data sets, we introduce Downpour SGD, a variant of asynchronous stochastic gradient descent that uses multiple replicas of a single DistBelief model. The basic approach is as follows: We divide the training data into a number of subsets and run a copy of the model on each of these subsets. The models communicate updates through a centralized parameter server, which keeps the current state of all parameters for the model, sharded across many machines (e.g., if we have 10 parameter server shards, each shard is responsible for storing and applying updates to 1/10th of the model parameters) (Figure 2). This approach is asynchronous in two distinct aspects: the model replicas run independently of each other, and the parameter server shards also run independently of one another.

**Comment by iamtrask** [comment]
irrelevant given new promise architecture
